src/common/ProducerConsumerQueue.h

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2015 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// @author Bo Hu (bhu@fb.com)
+// @author Jordan DeLong (delong.j@fb.com)
+// @author shawn chen(chen.xiaowei@h3c.com)
+
+#ifndef PRODUCER_CONSUMER_QUEUE_H_
+#define PRODUCER_CONSUMER_QUEUE_H_
+
+#include <atomic>
+#include "include/assert.h"
+
+#include <cstdlib>
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
+
+namespace folly {
+
+/*
+ * ProducerConsumerQueue is a one producer and one consumer queue
+ * without locks.
+ */
```

Comment:
So this is a single producer and single consumer lock-free queue, I guess this doesn't meet filejournal queue's case. It will be multi producer(pg thread).